---
title: "Description of snps read in with .vcf file and VariantAnnotation"
author: "Ann Von Holle"
date: '`r format(Sys.time(), "%d %B, %Y")`'
output:
  html_document:
    css: ../../style.css
    toc: yes
    toc_depth: '3'
    toc_float: no
    number_sections: yes
    xnos-section-numbers: yes
    pandoc_args:
    - --filter
    - pandoc-fignos
  pdf_document:
    number_sections: yes
    pandoc_args:
    - --filter
    - C:/Python27/Scripts/pandoc-eqnos.exe
    toc: yes
    toc_depth: 3
header-includes: \usepackage{hyperref}\hypersetup{colorlinks=true, linkcolor=blue,filecolor=magenta,urlcolor=cyan}
  \usepackage{amsmath}
---

<!-- NOTE: This program code is copied from look-vcf-snps.Rmd but cleaned up-->

```{r setup, include=FALSE}
knitr::opts_chunk$set(fig.width=12, fig.height=8,
                      echo=T, warning=FALSE,
                      message=T, 
                      results='markup',
                      comment=NA)
options(width = 150)
```

```{r }

#install.packages("PredictABEL", 
#                 repos="http://cran.r-project.org", 
#                 lib="/proj/epi/CVDGeneNas/avonholle/Rlibs", 
#                 dependencies=TRUE)

# see https://stackoverflow.com/questions/15170399/changing-r-default-library-path-using-libpaths-in-rprofile-site-fails-to-work
.libPaths( c( .libPaths(), "/proj/epi/CVDGeneNas/avonholle/Rlibs") )

library(PredictABEL, lib.loc="/proj/epi/CVDGeneNas/avonholle/Rlibs")

require(knitr)
#library(VariantAnnotation)
library(data.table)
#require(kableExtra)
```


```{r}

# from read-vcf-snps.R
# Look at Read of SLCS vcf files
# ---------------------------------------

# TO DO: need to go back and put this program in m3 folder
setwd("/proj/epi/CVDGeneNas/avonholle/ms-d3")
load("datsnp.Rdata") # dat.snp is a list of objects, first object is chr,
# second is data frame of snps
# and third item is the snp coding info: according to the documentation for variantannotation
# package, In the map DataFrame, allele.1 represents the reference allele and allele.2 is the alternate
# allele.

class(dat.snp)
#dat.snp[[1]][[2]]
head(dat.snp[[8]][[2]]) # test. 2nd element is the genotyping itself
head(dat.snp[[8]][[3]]) # test. third element is genotyping info. allele.1 is the reference allele, allele.2 is the risk allele
```



```{r}
# Extract out the info data from vcf file, which includes the maf. 4th element from list created in read-vcf-snps.R

info.dat = lapply(dat.snp, function(x) data.frame(x[4]))
info.dat[[1]]
info.dat.df = do.call(rbind.data.frame, info.dat)
info.dat.df

setwd("/proj/epi/CVDGeneNas/avonholle/ms-d3/data-prep")
save(info.dat.df, file="info-dat.Rdata")

```


```{r}
# Extract out genotype info for the samples across all chromosomes.
# -----------------------------------------------------------------

# Extract out info from snps, 2nd element
snp.dat = sapply(dat.snp, function(x) data.frame(x[2]))

snp.dat.df = do.call(cbind.data.frame, snp.dat)
dim(snp.dat.df)  # 881 obs and 207 snps for combined generalized + HL
snp.dat.df$id = substr(rownames(snp.dat.df), 3, length(rownames(snp.dat.df)))
snp.dat.df[1:5,c(ncol(snp.dat.df)-5, ncol(snp.dat.df))]

# Handle data so the ids match the phenotype data and snps in columns have rsids
# ----------------------------------------------------------------------------

nms = names(snp.dat.df); head(nms); length(unique(nms)) # 208 unique snps

# Extract out pos from columns of snp.dat.df
pos.1 = as.numeric(sapply(strsplit(nms, "\\."), "[", 2))
head(pos.1) # for some reason there are duplicaties (columns 66 and 67)

df.pos.1 = data.frame(Pos_hg19=pos.1, orig.ord = 1:length(pos.1))
head(df.pos.1)

# get rsid linked to hg19 position
# before, only had HL snps. Now, using combo of European generalizable to HL and HL.
# from tot.rsid data frame in 'create-list-snps-for-vcf.R'

# load the snps and effect sizes from spreadsheets, df.allele

load("vars.Rdata") # has tot.rsid data frame with chr and hg19 position from create-list-snps-for-vcf.R
# and df.allele -- data from studies
summary(df.allele$eur)
head(df.allele[c("eur", "Author", "rsid")])
```

```{r}
# get list of position numbers and rsid so I can put rsid on to position number 
# in column names from vcf file
link.pos.rsid = out.2[c("Pos_hg19", "rsid")] # total snps, 208
nrow(link.pos.rsid)
length(unique(link.pos.rsid$Pos_hg19)) # 208

# merge the rsid onto position number
nrow(df.pos.1) # 208
nrow(link.pos.rsid)
link.1 = merge(df.pos.1, 
               link.pos.rsid, 
               by="Pos_hg19", all.y=T)
dim(link.1) #208 unique snps
head(link.1)
# link.1 has data frame with rsid and position
dim(snp.dat.df)
head(snp.dat.df[,207:208])

# rename columns of snp dat by rsid and not position
snp.dat.df2 = snp.dat.df

#colnames(snp.dat.df2) = link.1$rsid
#snp.dat.df2[1:5,1:5]
dim(snp.dat.df2) # 881 by 208

# see https://stackoverflow.com/questions/26391921/how-to-convert-entire-dataframe-to-numeric-while-preserving-decimals

sapply(snp.dat.df2, class)
snp.dat.df2[,208]



# get phenotype id merged back on to data
# ------------------------------------------

ids = read.csv("IDsToMatch.csv", header=T) # ids from Anne that have the .ped/.fam file ids
colnames(ids) = tolower(colnames(ids))
ids$id_v1 = ids$id # make an id to merge onto the phenotype file
head(ids)

nrow(snp.dat.df2) #881

# have to fix id -- can't convert to numeric class in current state
id2 = with(snp.dat.df2, {
  ifelse(is.na(as.numeric(as.character(id)))==T,
         sapply(strsplit(as.character(id), "_"), "[", 2),
         as.numeric(as.character(id)))
})
id2

# this will now have phenotype id for merging with phenotype data in further analyses
colnames(snp.dat.df2)[208]
colnames(snp.dat.df2)[208] = "iid"
snp.dat.df2$iid = as.numeric(as.character(id2))

summary(snp.dat.df2$iid)
head(ids$iid)
nrow(ids) # 896


# Need to replace colnames with rsid
clean.snp.dat2 = merge(x=snp.dat.df2, 
                       y=ids[c("iid", "id")],
                       by="iid",
                       all.x=T)

head(colnames(clean.snp.dat2))
dim(snp.dat.df2)
dim(ids)
dim(clean.snp.dat2) # 881 by 209
tail(colnames(clean.snp.dat2))
head(colnames(clean.snp.dat2))

# column names 2-208 is "X.", chr # and position
cn = colnames(clean.snp.dat2)[2:208]
length(cn) # 207
head(cn)
cn.pos = as.numeric(sapply(strsplit(cn, "\\."), "[", 2))
head(cn.pos)
head(link.pos.rsid)
tail(link.pos.rsid)
dim(link.pos.rsid) # 208

cn.pos.df = data.frame(Pos_hg19=cn.pos, orig.vcf.order=1:length(cn.pos))
dim(cn.pos.df)
head(cn.pos.df)
tail(cn.pos.df) # 207

dim(link.pos.rsid)
cn.merge = merge(cn.pos.df, 
                 link.pos.rsid,
                 by="Pos_hg19",
                 all.x=T)
dim(cn.merge) # 207 by 3
head(cn.merge[order(cn.merge$orig.vcf.order),])
tail(cn.merge[order(cn.merge$orig.vcf.order),])
# now add rsid as column names for clean.snp.dat2 instead of pos

ncol1 = ncol(clean.snp.dat2); ncol1 # 209
clean.snp.dat2. = clean.snp.dat2
length(cn.merge$rsid)
length(2:ncol1-1) # 208
head(colnames(clean.snp.dat2.))
tail(colnames(clean.snp.dat2.))
cn.merge = cn.merge[order(cn.merge$orig.vcf.order),]
head(cn.merge[1:(ncol1-2),])
tail(cn.merge[1:(ncol1-2),])

length(colnames(clean.snp.dat2.)[2:(ncol1-1)]) # 207
length(cn.merge$rsid[1:(ncol1-2)]) # 207

colnames(clean.snp.dat2.)[2:(ncol1-1)] = cn.merge$rsid[1:(ncol1-2)]
head(clean.snp.dat2.[,1:5])
head(clean.snp.dat2.[,(ncol1-5):ncol1])
head(clean.snp.dat2[,(ncol1-5):ncol1]) # check

dim(clean.snp.dat2.) # 881 by 209 (881 poeple by 208 snps)
clean.snp.dat2.[875:881, 207:209] # check
clean.snp.dat2 = clean.snp.dat2.
```




```{r}

# read in pcs from /nas02/depts/epi/Genetic_Data_Center/chile/phenotypes/

setwd("/nas02/depts/epi/Genetic_Data_Center/chile/phenotypes/")
pcs = read.table("PCs_chileonly.txt")
head(pcs)
# NOTE: pcs mislabeled row headers.

summary(pcs)
dim(pcs) # 881 by 21
pcs$id = as.numeric(as.character(rownames(pcs)))
```


```{r}
# add pcs to clean.snp.dat2.
dim(clean.snp.dat2.)

head(clean.snp.dat2.$iid)
head(pcs$id)
clean.snp.dat2. = merge(x=clean.snp.dat2.,
                       y=pcs,
                       by.x="iid",
                       by.y="id",
                       all.x=T)

dim(clean.snp.dat2.)
head(clean.snp.dat2.[,(ncol(clean.snp.dat2.)-5):ncol(clean.snp.dat2.)])
tail(clean.snp.dat2.[,(ncol(clean.snp.dat2.)-5):ncol(clean.snp.dat2.)])

clean.snp.dat2 = clean.snp.dat2.
dim(clean.snp.dat2)
names(clean.snp.dat2)

setwd("/proj/epi/CVDGeneNas/avonholle/ms-d3/data-prep")
save(clean.snp.dat2, file="lipid-snp-dat.Rdata")

```

```{r}
# load in rsid for each lipid trait from power-calcs-ind-assoc.Rmd at
# ~\GitHub\unc-dissertation-markdown-p2\includes\scripts\power\aim3
load(file="sig-snps.Rdata") # has snp.sig with the 10 snps for further analysis and sort., the full list of snps for risk scores.

hdl.snps = sort.[sort.$Trait=="HDL", c("Trait", "rsid") ]
hdl.snp.list = dput(hdl.snps$rsid)

ldl.snps = sort.[sort.$Trait=="LDL", c("Trait", "rsid") ]
ldl.snp.list = dput(ldl.snps$rsid)

tg.snps = sort.[sort.$Trait=="TG", c("Trait", "rsid") ]
tg.snp.list = dput(tg.snps$rsid)

```


```{r}
# ---------------- END OF PART CREATING DATA FRAME OF SNP VALUES FOR BY PERSON ID -----

```



```{r, make-ungrs2, eval=T, echo=TRUE, results='markup'}

# use PredictABEL to create risk scores instead of manually as originally done.
# variants are from columns 2:208 (207 snps) in clean.snp.dat2.

# Get lipid traits to create beta for weights.

load(file="phen.Rdata") # get phenotype data from local drive. made in get-phen-data.Rmd file. data frames include: sub.phen, re.wt, re.wfl, re.ht
names(sub.phen)
lipid.t = sub.phen[c("id", "HDLchlstrl", "LDLChlstrl", "trg.mg.dL.")]
dim(lipid.t) # 596 by 4
names(lipid.t) = c("id", "hdl", "ldl", "tg")
lipid.t$log.tg = log(lipid.t$tg)
summary(lipid.t)

# merge lipid pheno to lipid traits to get betas
dim(clean.snp.dat2) # check
dim(lipid.t) # check
df.1 = merge(clean.snp.dat2,
             lipid.t, by="id")
dim(df.1) #check

# Export df.1 for descriptive table
# =================================

save(df.1, file="all-snps.Rdata")

# HDL
# ============================
# pull out snps for HDL trait

hdl.snps = df.1[c("id", hdl.snp.list)] # this data frame should have id in first column and rest are hdl-related snps
str(hdl.snps)
dim(hdl.snps)

# get beta for snps
beta.vec.h = sapply(hdl.snps[hdl.snp.list], function(x) coef(summary(lm(df.1$hdl ~ x)))[2,1])
beta.vec.h

# compute unweighted risk scores 
un.grs.hdl <- riskScore(weights=beta.vec.h, 
                        data=hdl.snps,
                        cGenPreds=hdl.snp.list,
                        Type="unweighted")
summary(un.grs.hdl)
length(un.grs.hdl)


# LDL
# ============================
# pull out snps for LDL trait

ldl.snp.list.2 = colnames(df.1)[colnames(df.1) %in% ldl.snp.list]
head(ldl.snp.list.2)

ldl.snps = df.1[c("id", ldl.snp.list2)] # this data frame should have id in first column and rest are ldl-related snps
str(ldl.snps)
dim(ldl.snps)

# get beta for snps
head(ldl.snps)
beta.vec.l = sapply(ldl.snps[ldl.snp.list.2],
                    
                    function(x) tryCatch(coef(summary(lm(df.1$ldl ~ x)))[2,1], error=function(err) NA))
beta.vec.l

# There are some missing values for some snps. omit
beta.vec.l.cc = beta.vec.l[complete.cases(beta.vec.l)==T]
head(beta.vec.l.cc)
beta.vec.l[complete.cases(beta.vec.l)==F] # snps not in data frame
names.nomiss = names(beta.vec.l.cc)

# compute unweighted risk scores 
un.grs.ldl <- riskScore(weights=beta.vec.l.cc, 
                        data=ldl.snps,
                        cGenPreds=names.nomiss,
                        Type="unweighted")
summary(un.grs.ldl)
length(un.grs.ldl)

# tg
# ============================
# pull out snps for tg trait

tg.snps = df.1[c("id", tg.snp.list)] # this data frame should have id in first column and rest are tg-related snps
str(tg.snps)
dim(tg.snps)

# get beta for snps
beta.vec.t = sapply(tg.snps[tg.snp.list], function(x) coef(summary(lm(df.1$tg ~ x)))[2,1])
beta.vec.t

# compute unweighted risk scores 
un.grs.tg <- riskScore(weights=beta.vec.t, 
                        data=tg.snps,
                        cGenPreds=tg.snp.list,
                        Type="unweighted")
summary(un.grs.tg)
length(un.grs.tg)

df.grs = data.frame(un.grs.hdl = un.grs.hdl,
                    un.grs.ldl = un.grs.ldl,
                    un.grs.tg = un.grs.tg)

# save grs
save(df.grs,  file="grs.Rdata")

```


```{r}
# checking coding of alleles


# Extract out info from map, 3rd element from each list
code.info = lapply(dat.snp, function(x) data.frame(x[3]))
class(code.info)
code.info.df = do.call(rbind.data.frame, code.info)
head(code.info.df)
dim(code.info.df) # 207 by 4


# extract out chr and position from snp.names column
code.info.df$Chr = paste0(sapply(strsplit(code.info.df$snp.names, ":"), "[", 1))
code.info.df$pos = paste0(sapply(strsplit(code.info.df$snp.names, ":"), "[", 2))
head(code.info.df)
dim(code.info.df) # 207 by 6


# having some difficulty with 'asis' class, see 
# see https://stackoverflow.com/questions/32112086/convert-asis-to-numeric-separated-by-coma-in-data-frame
code.info.df$allele.2.char <- vapply(code.info.df$allele.2, 
                                  function(x) paste(x, collapse = ","), character(1L))
code.info.df$allele.1.char <- vapply(code.info.df$allele.1, 
                                  function(x) paste(x, collapse = ","), character(1L))
class(code.info.df$allele.2.char)
head(code.info.df)

code.info.df2 = merge(code.info.df, 
                      df.allele[c("rsid", "Pos_hg19", "Chr")],
                      by.x = c("Chr", "pos"), 
                      by.y = c("Chr", "Pos_hg19")) # add rsid so I can get maf by rsid

code.info.df$gt = with(code.info.df, paste0(allele.2.char, "/", allele.1.char))
save(code.info.df2, file="vcf-map.Rdata") # save to file for checking

# allele.2 is the alternate allele and allele.1 is the reference allele
# in code.info.df


```

```{r}

# add maf from vcf file to this file
# from look-vcf-snps.Rmd

setwd("/proj/epi/CVDGeneNas/avonholle/ms-d3/data-prep")
load(file="info-dat.Rdata") # has info.dat.df data frame with maf and position # from vcf files
str(info.dat.df)

rownames(info.dat.df)
summary(info.dat.df$MAF)

# extract chr and position number from rownames
info.dat.df$Chr = paste0(sapply(strsplit(rownames(info.dat.df), ":"), "[", 1))
info.dat.df$pos = paste0(sapply(strsplit(rownames(info.dat.df), ":"), "[", 2))

head(info.dat.df)
dim(info.dat.df) # 207 unique snps by 6 columns of info
```




```{r}

# fix ids to phenotype ids
# get phenotype id merged back on to data
# ------------------------------------------
setwd("~/../Dropbox/unc.grad.school/my-papers/ms-201608-1/programs/kure-analysis")
ids = read.csv("IDsToMatch.csv", header=T) # ids from Anne that have the .ped/.fam file ids
colnames(ids) = tolower(colnames(ids))
ids$id_v1 = ids$id # make an id to merge onto the phenotype file
head(ids)
```

```{r, eval=F}

df.grs$iid = df.grs$id

df.grs = merge(df.grs,
               ids[c("iid", "id")], 
               by = "iid")
dim(df.grs)
df.grs = df.grs[c("id", "un.grs.hdl", "un.grs.ldl", "un.grs.tg")]
head(df.grs)
summary(df.grs)

save(df.grs, file="grs.Rdata")

```



```{r read}
# Code to read in select snps from vcf file
setwd("/proj/epi/CVDGeneNas/avonholle/ms-d3/data-prep")
load("lipid-snp-dat.Rdata") # has clean.snp.dat2 data frame with individual snp data
# from the 'look-vcf-snps.Rmd' program

str(clean.snp.dat2)

# select out the candidate snps from table 3.1 in dissertation proposal,
# has been revised as of 9/2017. Based on HL snsp and HL snps that generalize to 
# European snps.
# file is ~\GitHub\unc-dissertation-markdown\includes\scripts\power\aim3\power-calcs-ind-assoc.Rmd

select. = read.csv("sig-snp.csv")
head(select.)

select.rsid = select.$rsid
head(as.character(select.rsid))
length(select.rsid) # 10 snps

cn = colnames(clean.snp.dat2) %in% c("id", as.character(select.rsid))

colnames(clean.snp.dat2)[cn] # use these rsid in export-mplus-m3.Rmd file

pc.vars = c("plink_id", "PC1", "PC2", "PC3", "PC4")
sub.snp = clean.snp.dat2[, colnames(clean.snp.dat2) %in% c("id", pc.vars,
                                                           as.character(select.rsid))]

colnames(sub.snp)[which(colnames(sub.snp) %in% pc.vars)] = c("PC1", "PC2", "PC3", "PC4", "PC5")
# NOTE: the pc vars were mislabeled.

dim(sub.snp) # 881 by 16
head(sub.snp)
colnames(sub.snp)

```


```{r}

# get allele coding from vcf file used for sls genotyping
# ------------------------------------------------------------------
names(df.allele)
head(df.allele[c("rsid", "Pos_hg19", "Chr", "Effect allele", "Other allele", "EAF", "Beta (SE)", "Author", "beta.units", "Trait")])

# Extract out info from map, 3rd element from each list
code.info = lapply(dat.snp, function(x) data.frame(x[3]))
class(code.info)
code.info.df = do.call(rbind.data.frame, code.info)
head(code.info.df)
dim(code.info.df) # 207 by 4


# extract out chr and position from snp.names column
code.info.df$Chr = paste0(sapply(strsplit(code.info.df$snp.names, ":"), "[", 1))
code.info.df$pos = paste0(sapply(strsplit(code.info.df$snp.names, ":"), "[", 2))
head(code.info.df)
dim(code.info.df) # 207 by 6

code.info.df[code.info.df$pos %in% c('19830921', '116619073'),]

# having some difficulty with 'asis' class, see 
# see https://stackoverflow.com/questions/32112086/convert-asis-to-numeric-separated-by-coma-in-data-frame
code.info.df$allele.2.char <- vapply(code.info.df$allele.2, 
                                  function(x) paste(x, collapse = ","), character(1L))
code.info.df$allele.1.char <- vapply(code.info.df$allele.1, 
                                  function(x) paste(x, collapse = ","), character(1L))
class(code.info.df$allele.2.char)
head(code.info.df)

code.info.df2 = merge(code.info.df, 
                      df.allele[c("rsid", "Pos_hg19", "Chr")],
                      by.x = c("Chr", "pos"), 
                      by.y = c("Chr", "Pos_hg19")) # add rsid so I can get maf by rsid

code.info.df$gt = with(code.info.df, paste0(allele.2.char, "/", allele.1.char))
save(code.info.df2, file="vcf-map.Rdata") # save to file for checking
```

```{r, eval=T}

load(file="phen.Rdata") # get phenotype data from local drive. made in get-phen-data.Rmd file. data frames include: sub.phen, re.wt, re.wfl, re.ht

# merge the sitar random effects to the selected candidate snps
nrow(sub.snp)
colnames(sub.snp)
nrow(grs.dat)
nrow(sub.phen)
nrow(re.wt)

names(sub.snp)

sub.dat = merge(sub.snp, re.wt, by="id", all.x=T)
sub.dat = merge(sub.dat, re.ht, by="id", all.x=T)
sub.dat = merge(sub.dat, re.wfl, by="id", all.x=T)
sub.dat = merge(sub.dat, sub.phen, by="id", all.y=T)
nrow(sub.dat)
sub.dat = merge(sub.dat, grs.dat, by="id") #  data with phenotypes, n=653
# doesn't completely overlap with genetic data -- restricting data to n=523
sub.dat = sub.dat[sub.dat$milkcode %in% c(1,3),] # take out missing milkcode or milkcode=2, n=484

dim(sub.dat)

sub.dat = within(sub.dat, {
  mk2 = ifelse(milkcode==1, 1,
               ifelse(milkcode==3, 0, NA))
  gestage.ctr = scale(GestatnlAg, scale=F) # center gest age
  graffar.i.ctr = scale(graffar.index, scale=F) # center graffar
})

table(sub.dat$mk2) #check

nrow(sub.dat)
summary(sub.dat)

names(sub.dat)
table(sub.dat$milkcode)
table(sub.dat$Sex.f)

names(sub.dat)
str(sub.dat)
# export data for use in Mplus for second part of aim.
setwd("/proj/epi/CVDGeneNas/avonholle/ms-d3/data-prep")

# make a separate data frame for descriptive purposes 
all.dat = merge(clean.snp.dat2, sub.phen, by="id", all.y=T)

save(sub.dat, all.dat, file="subdat.Rdata")


```